src/search/plugins/plugin.h

```diff
@@ -1,103 +1,245 @@
+    const std::vector<PropertyInfo> &get_properties() const;
+    const std::vector<LanguageSupportInfo> &get_language_support() const;
+    const std::vector<NoteInfo> &get_notes() const;
+};
+
+
+template<typename Constructed>
+class FeatureWithDefault : public Feature {
+protected:
+    using Feature::Feature;
+    virtual std::shared_ptr<Constructed> create_component(const Options &options, const ConstructContext &) const {
+        return std::make_shared<Constructed>(options);
+    }
+};
+
+template<typename Constructed>
+class FeatureWithoutDefault : public Feature {
+protected:
+    using Feature::Feature;
+    virtual std::shared_ptr<Constructed> create_component(const Options &, const ConstructContext &) const = 0;
+};
+
+template<typename Constructed>
+using FeatureAuto = typename std::conditional<
+    std::is_constructible<Constructed, const Options &>::value,
+    FeatureWithDefault<Constructed>,
+    FeatureWithoutDefault<Constructed>>::type;
+
+template<typename Base, typename Constructed>
+class TypedFeature : public FeatureAuto<Constructed> {
+    using BasePtr = std::shared_ptr<Base>;
+    static_assert(std::is_base_of<Base, Constructed>::value,
+                  "Constructed must derive from Base");
+public:
+    TypedFeature(const std::string &key)
+        : FeatureAuto<Constructed>(TypeRegistry::instance()->get_type<BasePtr>(), key) {
+    }
 
+    Any construct(const Options &options, const ConstructContext &context) const override {
+        std::shared_ptr<Base> ptr = this->create_component(options, context);
+        return Any(ptr);
+    }
+};
+
+class Plugin {
+public:
+    Plugin();
+    virtual ~Plugin() = default;
+    Plugin(const Plugin &) = delete;
+    virtual std::shared_ptr<Feature> create_feature() const = 0;
+};
 
 template<typename T>
-class PluginTypePlugin {
+class FeaturePlugin : public Plugin {
 public:
-    PluginTypePlugin(
-        const std::string &type_name,
-        const std::string &documentation,
-        const std::string &predefinition_key = "",
-        const std::string &alias = "") {
-        using TPtr = std::shared_ptr<T>;
-        assert(!predefinition_key.empty() || alias.empty());
-        assert(!utils::startswith(predefinition_key, "--"));
-        assert(!utils::startswith(alias, "--"));
-
-        PredefinitionFunction predefinition_function = predefinition_key.empty() ?
-            nullptr : predefine_plugin<T>;
-
-        RawRegistry::instance()->insert_plugin_type_data(
-            std::type_index(typeid(TPtr)), type_name, documentation,
-            predefinition_key, alias, predefinition_function);
+    FeaturePlugin() : Plugin() {
+    }
+    virtual std::shared_ptr<Feature> create_feature() const override {
+        return std::make_shared<T>();
     }
+};
 
-    ~PluginTypePlugin() = default;
+/*
+  The type plugin class contains meta-information for a given
+  type of feature (e.g. "SearchEngine" or "MergeStrategyFactory").
+*/
```

Comment:
Good catch, the class indeed was called differently earlier. I fixed these comments but some ambiguity remains because we also have a class `FeatureType` that represents the C++ type of a feature. So for example for `Evaluator`, there is a `CategoryPlugin` and a `FeatureType` and many things that can be said about one could be said about the other.